core/sr-api-macros/src/impl_runtime_apis.rs

```diff
@@ -0,0 +1,530 @@
+/// The generated code includes decoding of the input arguments and encoding of the output.
+fn generate_impl_call(
+	signature: &MethodSig,
+	runtime: &Type,
+	input: &Ident,
+	impl_trait: &Path
+) -> Result<TokenStream> {
+	let mut pnames = Vec::new();
+	let mut ptypes = Vec::new();
+	let mut generated_pattern_counter = 0;
+	for input in signature.decl.inputs.iter() {
+		match input {
+			FnArg::Captured(arg) => {
+				match &arg.ty {
+					Type::Reference(_) => {
+						return Err(
+							Error::new(
+								arg.ty.span(),
+								"No type references are allowed in the api traits!"
+							)
+						)
+					},
+					_ => {},
+				}
+
+				pnames.push(
+					generate_unique_pattern(arg.pat.clone(), &mut generated_pattern_counter)
+				);
+				ptypes.push(&arg.ty);
+			},
+			_ => {
+				return Err(
+					Error::new(
+						input.span(),
+						"Only function arguments with the following \
+						pattern are accepted: `name: type`!"
+					)
+				)
+			}
+		}
+	}
+
+	let c = generate_crate_access(HIDDEN_INCLUDES_ID);
+	let c_iter = iter::repeat(&c);
+	let fn_name = &signature.ident;
+	let fn_name_str = iter::repeat(fn_name.to_string());
+	let input = iter::repeat(input);
+	let pnames2 = pnames.clone();
+
+	Ok(
+		quote!(
+			#(
+				let #pnames : #ptypes = match #c_iter::runtime_api::Decode::decode(&mut #input) {
+					Some(input) => input,
+					None => panic!("Bad input data provided to {}", #fn_name_str),
+				};
+			)*
+
+			let output = <#runtime as #impl_trait>::#fn_name(#( #pnames2 ),*);
+			#c::runtime_api::Encode::encode(&output)
+		).into()
+	)
+}
+
+/// Extract the trait that is implemented in the given `ItemImpl`.
+fn extract_impl_trait<'a>(impl_: &'a ItemImpl) -> Result<&'a Path> {
+	impl_.trait_.as_ref().map(|v| &v.1).ok_or_else(
+		|| Error::new(impl_.span(), "Only implementation of traits are supported!")
+	).and_then(|p| {
+		if p.segments.len() > 1 {
+			Ok(p)
+		} else {
+			Err(
+				Error::new(
+					p.span(),
+					"The implemented trait has to be referenced with a path, \
+					e.g. `impl client::Core for Runtime`."
+				)
+			)
+		}
+	})
+}
+
+/// Extracts the runtime block identifier.
+fn extract_runtime_block_ident(trait_: &Path) -> Result<&TypePath> {
+	let span = trait_.span();
+	let segment = trait_
+		.segments
+		.last()
+		.ok_or_else(
+			|| Error::new(span, "Empty path not supported")
+		)?;
+	let generics = segment.value();
+
+	if generics.arguments.is_empty() {
+		let span = trait_.segments.last().as_ref().unwrap().value().span();
+		Err(Error::new(span, "Missing `Block` generic parameter."))
+	} else {
+		match &generics.arguments {
+			PathArguments::AngleBracketed(ref args) => {
+				args.args.first().and_then(|v| match v.value() {
+					GenericArgument::Type(Type::Path(block)) => Some(block),
+					_ => None
+				}).ok_or_else(|| Error::new(args.span(), "Missing `Block` generic parameter."))
+			},
+			PathArguments::None => unreachable!(),
```

Comment:
why not have the code within the `if ...is_emtpy()` here instead and remove one indentation level? Is there a difference between `.is_empty()` and `==PathArguments::None`?